java-util/src/main/java/io/druid/java/util/common/guava/ExecuteWhenDoneYielder.java

```diff
@@ -56,9 +56,10 @@ public boolean isDone()
   @Override
   public void close() throws IOException
   {
-    if (isDone()) {
+    boolean done = isDone();
+    baseYielder.close();
```

Comment:
Not executing effect if exception is thrown in `close()` was aligned with behaviour in `accumulate()`, where `exec.execute(effect)` was just called after `subSequence.accumulate()`, without `finally`. However it seems reasonable to me to execute effect if exception is thrown in `close()`. To distinguish between exception thrown in `close()` and during the processing (before the sequence "is done") I had to use `YieldingSequenceBase`.

Also this whole thing of not executing the effect if the subsequence is not done is questionable. We use `Sequences.withEffect()` in two places in production. It seems to me that it makes sense to execute the effect in those places even if the subsequence is not "done" (especially considering that Sequences.limit() makes the sequence never done). What do you think?